docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,500 @@
+
+```csharp
+while (reader.Read())
+{
+  switch (reader.TokenType)
+  {
+    // ...
+    ReadOnlySpan<byte> jsonElement = reader.HasValueSequence ?
+                reader.ValueSequence.ToArray() :
+                reader.ValueSpan;
+    // ...
+  }
+}
+```
+
+### Read null values
+
+Code that uses the `Newtonsoft.Json` reader can compare a value token type of String to null, as in the following example:
+
+```csharp
+public static string ReadAsString(this JsonTextReader reader)
+{
+    reader.Read();
+
+    if (reader.TokenType != JsonToken.String)
+    {
+        throw new InvalidDataException();
+    }
+
+    return reader.Value?.ToString();
+}
+```
+
+This code doesn't work correctly in <xref:System.Text.Json>, which considers the `null` literal to be its own token type. When using `Utf8JsonReader`, check for the `JsonTokenType.Null` token type, as shown in the following example:
+
+```csharp
+public static string ReadAsString(this ref Utf8JsonReader reader)
+{
+    reader.Read();
+
+    if (reader.TokenType == JsonTokenType.Null)
+    {
+        return null;
+    }
+
+    if (reader.TokenType != JsonTokenType.String)
+    {
+        throw new InvalidDataException();
+    }
+
+    return reader.GetString();
+}
+```
+
+### Multi-targeting
+
+If you need to continue to use `Newtonsoft.Json` for certain target frameworks, you can multi-target and have two implementations. However, this is not trivial and would require some `#ifdefs` and source duplication. One way to share as much code as possible is to create a `ref struct` wrapper around `Utf8JsonReader` and `Newtonsoft.Json` `JsonTextReader`. This wrapper would unify the public surface area while isolating the behavioral differences. This lets you isolate the changes mainly to the construction of the type, along with passing the new type around by reference. This is the pattern that the [.NET Core installer](https://github.com/dotnet/runtime/tree/master/src/installer) follows:
+
+* [UnifiedJsonReader.JsonTextReader.cs](https://github.com/dotnet/runtime/blob/35fe82c2e90fa345d0b6ad243c14da193fb123fd/src/installer/managed/Microsoft.Extensions.DependencyModel/UnifiedJsonReader.JsonTextReader.cs)
+* [UnifiedJsonReader.Utf8JsonReader.cs](https://github.com/dotnet/runtime/blob/35fe82c2e90fa345d0b6ad243c14da193fb123fd/src/installer/managed/Microsoft.Extensions.DependencyModel/UnifiedJsonReader.Utf8JsonReader.cs)
+
+## Utf8JsonWriter
+
+<xref:System.Text.Json.Utf8JsonWriter?displayProperty=fullName> is a high-performance way to write UTF-8 encoded JSON text from common .NET types like `String`, `Int32`, and `DateTime`. The writer is a low-level type that can be used to build custom serializers.
+
+The following sections explain recommended programming patterns for using `Utf8JsonWriter`.
+
+### Write with UTF-8 text
```

Comment:
This also applies to folks who want custom escape behavior and use the [`WritePropertyName(string name, bool escape)`](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonWriter_WritePropertyName_1.htm) API.

`System.Text.Json` doesn't let you disable escaping while writing a string. However, you could pass in your own custom `JavascriptEncoder` as an option to the writer, or create your own `JsonEncodedText` which uses your `JavascriptEncoder` do to the escaping and then write the `JsonEncodedText` instead of the string.